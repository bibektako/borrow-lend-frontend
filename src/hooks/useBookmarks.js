import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query';
import { useContext } from 'react';
import { AuthContext } from '../auth/Authprovider';
import { toast } from 'react-hot-toast';

import {
  getBookmarksApi,
  addBookmarkApi,
  removeBookmarkApi,
} from '../api/authAPI'; // Make sure this path is correct

export const useBookmarks = () => {
  const { user } = useContext(AuthContext);
  const queryClient = useQueryClient();

  const { data: bookmarks = [], isLoading } = useQuery({
    queryKey: ['bookmarks', user?._id],
    queryFn: async () => {
      const response = await getBookmarksApi();
      return response.data.data; 
    },
    enabled: !!user,
  });

  const isBookmarked = (itemId) => {
    return bookmarks.some((bookmark) => bookmark._id === itemId);
  };

  const { mutate: toggleBookmark, isPending: isToggling } = useMutation({
    mutationFn: async ({ itemId, isCurrentlyBookmarked }) => {
      if (isCurrentlyBookmarked) {
        return removeBookmarkApi(itemId);
      } else {
        return addBookmarkApi(itemId);
      }
    },
    onSuccess: (response, variables) => {
      const message = variables.isCurrentlyBookmarked
        ? "Removed from wishlist"
        : "Added to wishlist";
      toast.success(message);

      queryClient.invalidateQueries({ queryKey: ['bookmarks', user?._id] });
    },
    onError: (error) => {
      toast.error(error.response?.data?.message || "Could not update wishlist.");
    },
  });

  const handleToggleBookmark = (itemId) => {
    const isCurrentlyBookmarked = isBookmarked(itemId);
    toggleBookmark({ itemId, isCurrentlyBookmarked });
  };

  return {
    bookmarks,
    isLoading,
    toggleBookmark: handleToggleBookmark, 
    isToggling,
    isBookmarked,
    bookmarkCount: bookmarks.length,
  };
};