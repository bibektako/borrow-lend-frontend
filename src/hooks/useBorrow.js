import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import {
  createBorrowRequestService,
  getBorrowRequestsService,
  updateBorrowRequestService,
} from "../services/borrowService";
import { toast } from "react-toastify";
import { useContext } from "react";
import { AuthContext } from "../auth/Authprovider";

export const useBorrowRequests = () => {
  const { token } = useContext(AuthContext);
  const { data, isLoading, error, isError, refetch} = useQuery({
    queryKey: ["borrow-requests"],
    queryFn: () => getBorrowRequestsService(token),
    enabled: !!token,
  });
  return {
    requests: data?.data || [],
    isLoading,
    error,
    isError,
    refetch
  };
};

export const useCreateBorrowRequest = () => {
  const queryClient = useQueryClient();
  const { token } = useContext(AuthContext);

  return useMutation({
    mutationFn: (itemId) => createBorrowRequestService(itemId, token),
    onSuccess: (data) => {
      toast.success(data?.message || "Borrow request sent successfully!");
      queryClient.invalidateQueries({ queryKey: ["items"] });
      queryClient.invalidateQueries({ queryKey: ["item"] });
      queryClient.invalidateQueries({ queryKey: ["borrow-requests"] });
    },
    onError: (error) => {
      toast.error(error?.message || "Failed to send borrow request.");
    },
  });
};

export const useUpdateBorrowRequest = () => {
  const queryClient = useQueryClient();
  const { token } = useContext(AuthContext);

  return useMutation({
    mutationFn: ({ requestId, status }) =>
      updateBorrowRequestService(requestId, status, token),
    onSuccess: (data) => {
      toast.success(data?.message || "Request updated!");
      queryClient.invalidateQueries({ queryKey: ["borrow-requests"] });
    },
    onError: (error) => {
      toast.error(error?.message || "Failed to update request.");
    },
  });
};
