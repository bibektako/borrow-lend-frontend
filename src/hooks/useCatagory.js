import { useQuery } from "@tanstack/react-query";

import { getAllCategoryService } from "../services/admin/categoryService";


export const useCategories = () => {
  const {
    data: categoryData, // We get the raw response here, let's call it categoryData
    isLoading,
    isError,
    error,
  } = useQuery({
    queryKey: ["categories"],
    
    queryFn: getAllCategoryService,
    
    staleTime: 1000 * 60 * 15,
  });

  const categories = categoryData?.data || [];

  return {
    categories, // The clean array of category objects
    isLoading,  // True while fetching
    isError,    // True if an error occurred
    error,      // The error object
  };
};