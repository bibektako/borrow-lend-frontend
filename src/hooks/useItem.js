import { useQuery, useMutation, useQueryClient } from "@tanstack/react-query";
import {
  getItemsService,
  getItemByIdService,
  getMyItemsService,
  createItemService,
  updateItemService,
  deleteItemService,
} from "../services/itemService";
import { getAllCategoryService } from "../services/admin/categoryService";
import { toast } from "react-toastify";
import { useContext } from "react";
import { AuthContext } from "../auth/Authprovider";

export const useItems = (filters = {}) => {
  const { data, isLoading, error, isError } = useQuery({
    queryKey: ["items", filters],
    queryFn: () => getItemsService(filters),
  });

  return {
    items: data?.data || [],
    isLoading,
    error,
    isError,
  };
};

export const useMyItems = () => {
  const { data, isLoading, error, isError } = useQuery({
    queryKey: ["my-items"],
    queryFn: getMyItemsService,
  });
  return {
    items: data?.data || [],
    isLoading,
    error,
    isError,
  };
};
export const useCategories = () => {
    const { data, isLoading, error } = useQuery({
        queryKey: ["categories"],
        queryFn: getAllCategoryService,
    });
    return {
        categories: data?.data?.categories || [],
        isLoadingCategories: isLoading,
        errorCategories: error,
    };
};

export const useItemById = (id) => {
  const { data, isLoading, error, isError } = useQuery({
    queryKey: ["item", id],
    queryFn: () => getItemByIdService(id),
    enabled: !!id,
  });

  return {
    item: data?.data || null,
    isLoading,
    error,
    isError,
  };
};

export const useCreateItem = () => {
  const queryClient = useQueryClient();
  const { token } = useContext(AuthContext);

  return useMutation({
    mutationFn: (formData) => createItemService(formData, token),
    onSuccess: (data) => {
      toast.success(data?.message || "Item created successfully!");
      queryClient.invalidateQueries({ queryKey: ["items"] });
    },
    onError: (error) => {
      toast.error(error?.message || "Failed to create item.");
    },
  });
};

export const useUpdateItem = () => {
  const queryClient = useQueryClient();
  const { token } = useContext(AuthContext);

  return useMutation({
    mutationFn: ({ id, formData }) => updateItemService(id, formData, token),
    onSuccess: (data, variables) => {
      toast.success(data?.message || "Item updated successfully!");

      queryClient.invalidateQueries({ queryKey: ["items"] });
      queryClient.invalidateQueries({ queryKey: ["item", variables.id] });
    },
    onError: (error) => {
      toast.error(error?.message || "Failed to update item.");
    },
  });
};

export const useDeleteItem = () => {
  const queryClient = useQueryClient();
  const { token } = useContext(AuthContext);

  return useMutation({
    mutationFn: (id) => deleteItemService(id, token),
    onSuccess: (data) => {
      toast.success(data?.message || "Item deleted successfully!");

      queryClient.invalidateQueries({ queryKey: ["items"] });
    },
    onError: (error) => {
      toast.error(error?.message || "Failed to delete item.");
    },
  });
};
