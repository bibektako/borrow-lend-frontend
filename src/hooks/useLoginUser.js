import { toast } from "react-toastify";
import { loginUserService } from "../services/authService";
import { useMutation } from "@tanstack/react-query";
import { AuthContext } from "../auth/Authprovider";
import { useContext } from "react";
export const useLoginUser =() => {
    const { login } = useContext(AuthContext)

    return useMutation(
        {
            mutationFn: loginUserService,
            mutationKey: ['login-key'],
            onSuccess: (data) => {
                login(data?.data, data?.token)
                toast.success(data?.message || "Login Success")
            },
            onError: (err)=>{
                toast.error(err?.message || "Login Failed")
            }
        }
    )

}