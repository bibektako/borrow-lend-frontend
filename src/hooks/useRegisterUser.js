import { useState } from "react"
import { registerUserService } from "../services/authService"

export const useRegisterUser = ()=>{
    const [isLoading, setIsLoading] = useState(false)
    const [error, setError] = useState(null)
    const [data, setData] = useState(null)

    const register = async (fromData) =>{
        setIsLoading(true)
        setError(null)
        setData(null)

        try {
            const response = await registerUserService(fromData)
            setData(response)
            return response
        } catch (err) {
            setError(err)
            return null
        }
        finally{
            setIsLoading(false)
        }
    }
    return {
        register,
        isLoading,
        data,
        error
    }
}