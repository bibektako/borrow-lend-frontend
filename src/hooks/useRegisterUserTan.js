import { useMutation } from "@tanstack/react-query";
import { registerUserService } from "../services/authService";
import { toast } from "react-toastify";

export const useRegisterUserTan = ()=>{
    return useMutation(
        {
            mutationFn: registerUserService,
            mutationKey: [ 'register'],
            onSuccess: (data) =>{
                toast.success(data.message || "Registration Successful")
            },
            onError: (err)=>{
                toast.error(err.message || "Registration Failed")
            }
        }
    )
}