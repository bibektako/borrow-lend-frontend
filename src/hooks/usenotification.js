import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import {
  createBorrowRequestService,
  getBorrowRequestsService,
  updateBorrowRequestService,
} from "../services/borrowService";
import { toast } from "react-toastify";
import { useContext } from "react";
import { AuthContext } from "../auth/Authprovider";

/**
 * Fetches all incoming and outgoing borrow requests for the logged-in user.
 */
export const useBorrowRequests = () => {
  const { token } = useContext(AuthContext);
  const { data, isLoading, error, isError, refetch } = useQuery({
    queryKey: ["borrow-requests"],
    queryFn: () => getBorrowRequestsService(token),
    enabled: !!token, // Only run the query if the user is logged in
  });

  return {
    requests: data?.data || [],
    isLoading,
    error,
    isError,
    refetch,
  };
};

/**
 * Provides a mutation function to create a new borrow request.
 */
export const useCreateBorrowRequest = () => {
  const queryClient = useQueryClient();
  const { token } = useContext(AuthContext);

  return useMutation({
    mutationFn: (itemId) => createBorrowRequestService(itemId, token),
    onSuccess: (data) => {
      toast.success(data?.message || "Borrow request sent successfully!");
      // Invalidate related queries to refetch data
      queryClient.invalidateQueries({ queryKey: ["items"] });
      queryClient.invalidateQueries({ queryKey: ["item"] });
      queryClient.invalidateQueries({ queryKey: ["borrow-requests"] });
    },
    onError: (error) => {
      toast.error(error?.message || "Failed to send borrow request.");
    },
  });
};

/**
 * Provides a mutation function to update a borrow request's status
 * with INSTANT UI feedback (Optimistic Update).
 */
export const useUpdateBorrowRequest = () => {
  const queryClient = useQueryClient();
  const { token } = useContext(AuthContext);

  return useMutation({
    mutationFn: ({ requestId, status }) =>
      updateBorrowRequestService(requestId, status, token),

    // This runs immediately on mutation, BEFORE the API call
    onMutate: async ({ requestId, status: newStatus }) => {
      // Cancel any ongoing refetches to prevent conflicts
      await queryClient.cancelQueries({ queryKey: ["borrow-requests"] });

      // Snapshot the previous state of the data
      const previousRequestsData = queryClient.getQueryData(["borrow-requests"]);

      // Optimistically update the cache
      queryClient.setQueryData(["borrow-requests"], (oldData) => {
        if (!oldData || !oldData.data) return oldData;
        const updatedRequests = oldData.data.map((request) =>
          request._id === requestId
            ? { ...request, status: newStatus } // Instantly change the status
            : request
        );
        return { ...oldData, data: updatedRequests };
      });

      // Return the snapshot to be used in case of an error
      return { previousRequestsData };
    },

    // If the API call fails, roll back to the snapshotted data
    onError: (error, variables, context) => {
      if (context?.previousRequestsData) {
        queryClient.setQueryData(["borrow-requests"], context.previousRequestsData);
      }
      toast.error(error?.message || "Action failed. Reverting change.");
    },

    onSettled: () => {
      queryClient.invalidateQueries({ queryKey: ["borrow-requests"] });
      queryClient.invalidateQueries({ queryKey: ['items'] });
      queryClient.invalidateQueries({ queryKey: ['item'] });
    },
  });
};