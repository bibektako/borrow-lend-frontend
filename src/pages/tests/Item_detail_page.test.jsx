// src/pages/__tests__/ItemDetailPage.test.jsx
import React from 'react';
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen } from '@testing-library/react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import ItemDetailPage from '../ItemsDetailPage';
import * as useItemHook from '../../hooks/useItem';

vi.mock('../../components/items/items detail/ImageGallery', () => ({
  default: () => <div data-testid="image-gallery">Image Gallery</div>,
}));
vi.mock('../../components/items/items detail/itemsInfo', () => ({
  default: () => <div data-testid="item-info">Item Info</div>,
}));
vi.mock('../../components/items/items detail/ownerInfo', () => ({
  default: () => <div data-testid="owner-info">Owner Info</div>,
}));
vi.mock('../../components/items/items detail/BookingPanel', () => ({
  default: () => <div data-testid="booking-panel">Booking Panel</div>,
}));
vi.mock('../../components/items/items detail/Reviews', () => ({
  default: () => <div data-testid="reviews">Reviews</div>,
}));

describe('ItemDetailPage', () => {
  const mockItem = {
    _id: '123',
    name: 'Test Item',
    imageUrls: ['url1', 'url2'],
    owner: { name: 'Owner' },
  };

  beforeEach(() => {
    vi.restoreAllMocks(); // Reset mocks between tests
  });

  it('shows loading state', () => {
    vi.spyOn(useItemHook, 'useItemById').mockReturnValue({
      item: null,
      isLoading: true,
      error: null,
    });

    render(
      <MemoryRouter initialEntries={['/items/123']}>
        <Routes>
          <Route path="/items/:id" element={<ItemDetailPage />} />
        </Routes>
      </MemoryRouter>
    );

    expect(screen.getByText(/loading item details/i)).toBeInTheDocument();
  });

  it('shows error state', () => {
    vi.spyOn(useItemHook, 'useItemById').mockReturnValue({
      item: null,
      isLoading: false,
      error: { message: 'Something went wrong' },
    });

    render(
      <MemoryRouter initialEntries={['/items/123']}>
        <Routes>
          <Route path="/items/:id" element={<ItemDetailPage />} />
        </Routes>
      </MemoryRouter>
    );

    expect(screen.getByText(/error/i)).toBeInTheDocument();
    expect(screen.getByText(/something went wrong/i)).toBeInTheDocument();
  });

  it('shows "item not found" message', () => {
    vi.spyOn(useItemHook, 'useItemById').mockReturnValue({
      item: null,
      isLoading: false,
      error: null,
    });

    render(
      <MemoryRouter initialEntries={['/items/123']}>
        <Routes>
          <Route path="/items/:id" element={<ItemDetailPage />} />
        </Routes>
      </MemoryRouter>
    );

    expect(screen.getByText(/item not found/i)).toBeInTheDocument();
  });

  it('renders all main components when item is available', () => {
    vi.spyOn(useItemHook, 'useItemById').mockReturnValue({
      item: mockItem,
      isLoading: false,
      error: null,
    });

    render(
      <MemoryRouter initialEntries={['/items/123']}>
        <Routes>
          <Route path="/items/:id" element={<ItemDetailPage />} />
        </Routes>
      </MemoryRouter>
    );

    expect(screen.getByTestId('image-gallery')).toBeInTheDocument();
    expect(screen.getByTestId('item-info')).toBeInTheDocument();
    expect(screen.getByTestId('owner-info')).toBeInTheDocument();
    expect(screen.getByTestId('booking-panel')).toBeInTheDocument();
    expect(screen.getByTestId('reviews')).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    vi.spyOn(useItemHook, 'useItemById').mockReturnValue({
      item: mockItem,
      isLoading: false,
      error: null,
    });

    const { asFragment } = render(
      <MemoryRouter initialEntries={['/items/123']}>
        <Routes>
          <Route path="/items/:id" element={<ItemDetailPage />} />
        </Routes>
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
