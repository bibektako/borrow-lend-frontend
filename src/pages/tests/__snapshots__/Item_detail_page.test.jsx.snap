// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ItemDetailPage > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="bg-gray-50 min-h-screen"
  >
    <section
      aria-atomic="false"
      aria-label="Notifications Alt+T"
      aria-live="polite"
      aria-relevant="additions text"
      class="Toastify"
    />
    <main
      class="container mx-auto px-4 py-8"
    >
      <div
        class="mb-6"
      >
        <a
          class="flex items-center gap-2 text-sm font-semibold text-gray-600 hover:text-gray-900"
          data-discover="true"
          href="/browse"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-arrow-left"
            fill="none"
            height="16"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m12 19-7-7 7-7"
            />
            <path
              d="M19 12H5"
            />
          </svg>
          Back to all items
        </a>
      </div>
      <div
        class="grid grid-cols-1 lg:grid-cols-3 gap-8 items-start"
      >
        <div
          class="lg:col-span-2 space-y-8"
        >
          <div
            data-testid="image-gallery"
          >
            Image Gallery
          </div>
          <div
            data-testid="item-info"
          >
            Item Info
          </div>
          <div
            data-testid="reviews"
          >
            Reviews
          </div>
        </div>
        <div
          class="lg:col-span-1 space-y-8"
        >
          <div
            data-testid="owner-info"
          >
            Owner Info
          </div>
          <div
            data-testid="booking-panel"
          >
            Booking Panel
          </div>
        </div>
      </div>
    </main>
  </div>
</DocumentFragment>
`;
